scene/gui/control.cpp

```diff
@@ -2978,30 +2977,21 @@ Control *Control::make_custom_tooltip(const String &p_text) const {
 // Base object overrides.
 
 void Control::add_child_notify(Node *p_child) {
-	Control *child_c = Object::cast_to<Control>(p_child);
-
-	if (child_c && child_c->data.theme.is_null() && (data.theme_owner || data.theme_owner_window)) {
-		_propagate_theme_changed(child_c, data.theme_owner, data.theme_owner_window); //need to propagate here, since many controls may require setting up stuff
-	}
-
-	Window *child_w = Object::cast_to<Window>(p_child);
+	if (data.theme_owner || data.theme_owner_window) {
+		// The added child won't be inheriting any theme items as it's been outside of the scene tree,
+		// so if this node doesn't have any non-default theme items for the child to inherit,
+		// then there's no need to propagate.
 
-	if (child_w && child_w->theme.is_null() && (data.theme_owner || data.theme_owner_window)) {
-		_propagate_theme_changed(child_w, data.theme_owner, data.theme_owner_window); //need to propagate here, since many controls may require setting up stuff
+		// We never need to notify here as if the node is inside the tree, it'll get notified
+		// from `NOTIFICATION_ENTER_TREE`, otherwise it's outside of the tree and we don't need to notify anyways.
```

Comment:
The first comment should probably go before the condition. I also find the explanation a bit heavy. Took me a few moments to understand what you want to imply here. Can you try simplifying it a bit?

Like, the point of the first one is that we only propagate if there is a theme owner, which indicates that we have some custom theme applied? And the point of the second is that `NOTIFICATION_ENTER_TREE` for that child will also notify about theme changes through a deferred call right after this code is executed?

Reply:
Fair enough, I was just trying to over-document rather than under-document